DOC-3498: tinymceai on-premises documentation (#4142)
tinymce/tinymce-docs: modules/ROOT/pages/tinymceai-on-premises-getting-started.adoc

= Getting started with TinyMCE AI On-Premise
:navtitle: Getting started
:description: Five-minute Docker Compose quick start for TinyMCE AI on-premises service
:keywords: AI, on-premises, getting started, docker, quick start

> This section produces a fully working setup (AI service, database, Redis, token server, and a browser editor) in roughly five minutes on any machine with Docker. This quick start validates the stack components before designing a production deployment. Production engineers should still review this section to understand the conceptual flow before continuing to xref:tinymceai-on-premises-production.adoc[the Production Deployment Guide].

Reviewer: ```suggestion
This section produces a fully working setup (AI service, database, Redis, token server, and a browser editor) in roughly five minutes on any machine with Docker. This quick start is designed to validate the stack components before moving to a production deployment. Production engineers should still review this section to understand the conceptual flow before continuing to xref:tinymceai-on-premises-production.adoc[the Production Deployment Guide].
```
maybe?? the wording tripped my brain up. not sure this is quite right either though. feel free to ignore

Author: Done — rewrote the opening to be a bullet-list summary of what the guide builds (MySQL, Redis, AI service, token server, browser page) followed by a clearer purpose statement. See ad5b4ed.

Author: Done — rewrote the opening as a bullet-list summary of what the guide builds (MySQL, Redis, AI service, token server, browser page) with a purpose statement and link to Production deployment. See e006b46.

== Five-minute demo with Docker Compose

=== Create the project folder

[source,bash]
----
mkdir tinymce-ai-onpremise && cd tinymce-ai-onpremise
----

=== Authenticate with the container registry

The service image lives at `registry.containers.tiny.cloud/ai-service-tiny`.

For Docker:

[source,bash]
----
docker login -u 'TINY_REGISTRY_USERNAME' https://registry.containers.tiny.cloud
# Docker prompts for the password; this avoids leaking it in shell history.
----

For Podman:

[source,bash]
----
podman login -u 'TINY_REGISTRY_USERNAME' registry.containers.tiny.cloud
----

> Replace `TINY_REGISTRY_USERNAME` with the username supplied by the Tiny account representative. If credentials have not been received, contact `support@tiny.cloud`.

Reviewer: I believe we plan to have this available in the customer portal eventually - just flagging that we'll want to make sure terminology lines up. Not a now problem, but for the future.

Author: Noted — flagged for future alignment when portal terminology is finalized. No action needed now.

Author: Noted — no action now, will align terminology when portal naming is finalized.

=== Pull the AI service image

[source,bash]
----
docker pull registry.containers.tiny.cloud/ai-service-tiny:latest
----

> For Podman, substitute `podman pull`. For production, pin a specific version tag (for example `:5.1.0`) rather than `:latest`.

Reviewer: why do we recommend pinning? feel like the why should be included.

Author: Good call — expanded the TIP to explain that the `:8` tag may pull MySQL versions with authentication plugins or SQL modes incompatible with the AI service schema migrations.

Author: Added — TIP now explains that `:8` may pull versions with incompatible authentication plugins or SQL modes. See e006b46.

> === Create `docker-compose.yml`

Reviewer: I'd like us to specify this is to start up the required data layer services - sql and redis. My initial assumption was that this was starting the AI service itself. 
The note in the following section about keeping AI + data layer separate allows fine grained control could maybe go here instead, to explain why we're doing these two first and not all 3 at once?

Author: Done — renamed the section to "Create docker-compose.yml (data layer)" and added an explanatory paragraph: "This compose file starts the data layer services (MySQL and Redis) that the AI service depends on. The AI service itself is started separately in the next step, which allows upgrading or reconfiguring it independently."

Author: Done — renamed section to "Create docker-compose.yml (data layer)" with an explanatory paragraph stating it starts MySQL and Redis only. The AI service is started separately in the next step. See e006b46.

Create the file with exactly the contents below. Indentation is two spaces, never tabs.

[source,yaml]
----
services:
  mysql:
    image: mysql:8.0
    environment:
      MYSQL_ROOT_PASSWORD: ${DB_PASSWORD:-changeme}
      MYSQL_DATABASE: ai_service
    ports:
      - "3306:3306"
    volumes:
      - mysql_data:/var/lib/mysql
    healthcheck:
      test: ["CMD", "mysqladmin", "ping", "-h", "localhost"]
      interval: 10s
      timeout: 5s
      retries: 5

  redis:
    image: redis:7
    ports:
      - "6379:6379"
    healthcheck:
      test: ["CMD", "redis-cli", "ping"]
      interval: 10s
      timeout: 5s
      retries: 5

volumes:
  mysql_data:
----

TIP: Pin `mysql:8.0`, not `mysql:8`. The `:8` tag points to the latest MySQL, which is incompatible with the AI service. See xref:tinymceai-on-premises-database.adoc#mysql-version-pinning[MySQL version pinning] for details.

> PostgreSQL is equally supported. See xref:tinymceai-on-premises-database.adoc[Database, Redis, and storage] for an equivalent compose file. Review the xref:tinymceai-on-premises-database.adoc#postgresql-schema-prerequisite[PostgreSQL schema prerequisite] before switching.

Reviewer: was thinking, if I was doing this quickstart I'd rather know at the start what I'm setting up, not discover as I go. could we put a bit at the start, maybe under "five minute quick start..." saying something like "this step by step guide will set up AI On Prem with a MySQL database (and any other choices we're making here)"? this paragraph could then potentially go after that instead of here too
we do define this more in this pr than the old one where I originally commented this, but it still stands

Author: Agreed — added an upfront component list at the top of the page showing exactly what gets set up (MySQL, Redis, AI service, token server, browser page) before any steps begin.

Author: Agreed — added a component list at the top (lines 6–12) listing everything the guide sets up, plus a Prerequisites section with software requirements. See e006b46.

[NOTE]
--
If any service in the stack needs to reach the host machine (for example a self-hosted Ollama running on the host), add an `extra_hosts` entry to the `ai-service` block in the compose file above:

[source,yaml]
----
extra_hosts:
  - "host.docker.internal:host-gateway"
----

Docker Desktop (macOS, Windows) and Podman 4{plus} auto-inject this alias. Native Linux Docker does not.
--

=== Create the `.env` file

[source,bash]
----
# --- Required: provided by Tiny ---
LICENSE_KEY=PASTE_SUPPLIED_LICENSE_KEY_HERE

> TINYMCE_API_KEY=PASTE_TINYMCE_API_KEY_HERE

Reviewer: My audit flagged that its weird we're requiring both a license key and API key here (on prem and cloud). Is this a one or the other situation, or are we overloading terms with new definitions?

Reviewer: Ah. License key is the Docker license key? Not a TinyMCE license key?

Author: Addressed — added a Credentials table to the Overview page explaining all three credentials (LICENSE_KEY, TINYMCE_API_KEY, license_key init option), their purposes, and when each is required. Also marked TINYMCE_API_KEY as optional in the .env file with a comment explaining it's only needed for CDN-loaded editor bundles.

Author: Addressed — the Overview page now has a dedicated Credentials section with a table explaining that LICENSE_KEY activates the AI service (provided by account rep), while TINYMCE_API_KEY authenticates the editor against cdn.tiny.cloud (from the dashboard). They are entirely separate credentials from different sources.

Author: Addressed — added an IMPORTANT block explaining the difference: LICENSE_KEY activates the AI service (from account rep), TINYMCE_API_KEY loads TinyMCE from CDN (not needed for self-hosted bundles). Also added a Credentials table to the Overview page. See e006b46.

Author: Clarified — LICENSE_KEY is the AI service activation key from the account representative (not the TinyMCE editor license key). Added an IMPORTANT block and a Credentials table on the Overview to make this unambiguous. See e006b46.

# --- Required: strong secret used to log into the Management Panel ---
MANAGEMENT_SECRET=REPLACE_WITH_STRONG_SECRET

# --- Required: database password (must match docker-compose.yml) ---
DB_PASSWORD=changeme

> 

Reviewer: inconsistent formating with the parts people need to change?

Reviewer: oh I see the db password is like that because its a callback to the docker compose. hm. ignore me if this makes sense as is

Author: Fixed — normalized all placeholder values to consistent `<angle-bracket-descriptor>` format (e.g. `<paste-license-key-here>`, `<replace-with-strong-secret>`, `<replace-with-db-password>`).

Author: No worries — the DB_PASSWORD placeholder is now also in the normalized format for consistency.

Author: Fixed — all placeholders now use consistent `<angle-bracket-descriptor>` format throughout the .env and code blocks. See e006b46.

Author: Makes sense as is — the DB_PASSWORD references the compose file variable. Normalized the format for consistency anyway.

# --- Required: at least one LLM provider key ---
OPENAI_API_KEY=sk-proj-PASTE_OPENAI_KEY_HERE
# ANTHROPIC_API_KEY=sk-ant-PASTE_ANTHROPIC_KEY_HERE
# GOOGLE_API_KEY=AIza-PASTE_GOOGLE_KEY_HERE

# --- Filled in after creating an environment (Step 7). Used by the token server, not the AI service. ---
AI_ENV_ID=
AI_API_SECRET=
----

IMPORTANT: `LICENSE_KEY` and `TINYMCE_API_KEY` are different credentials. `LICENSE_KEY` is the long string from the account representative. `TINYMCE_API_KEY` is the short string from the tiny.cloud dashboard.

=== Start MySQL and Redis

[source,bash]
----
docker compose up -d
----

Wait ~15 seconds for MySQL to initialize, then verify:

[source,bash]
----
docker compose ps
----

> Both containers should report `healthy` in the STATUS column. If MySQL still shows `starting`, wait another 10 seconds and re-run.

Reviewer: This might be me not knowing Docker but "both containers"? Is this mysql and redis?

Author: Fixed — changed to "Both data layer containers (MySQL and Redis) should report healthy..." to remove ambiguity.

Author: Fixed — changed to "Both data layer containers (MySQL and Redis) should report healthy..." See e006b46.

  <script>
    tinymce.init({
      selector: '#editor',
      plugins: 'tinymceai',
      toolbar: 'undo redo | blocks | bold italic | tinymceai-chat tinymceai-review tinymceai-quickactions',
      height: 500,
      tinymceai_service_url: '${AI_SERVICE_URL}',
      tinymceai_token_provider: () =>
        fetch('/api/ai-token', { method: 'POST' })
          .then(r => r.json())
          .then(data => ({ token: data.token }))
    });
  </script>
</body>
</html>`);
});

app.listen(PORT, () => {
  console.log('Editor:     http://localhost:' + PORT);
  console.log('Token API:  http://localhost:' + PORT + '/api/ai-token');
  console.log('AI Service: ' + AI_SERVICE_URL);
});
----
====

=== Install and run

[source,bash]
----
npm install
npm start
----

=== Open the demo

Open *http://localhost:3000* in a browser. The editor loads with the AI toolbar. Select text and try the AI features. Responses stream in real time from the chosen large language model (LLM) provider, processed entirely within the local infrastructure.

The TinyMCE AI on-premises service is now running.

> == Verifying the installation

Reviewer: this section is confusing to read because there's no breaks between the steps being checked. needs some subheadings / bold titles / at least a sentence about what each step is checking. otherwise the snippets for expected then next step flow into each other

<img width="1293" height="540" alt="Image" src="https://github.com/user-attachments/assets/a03efed4-ea07-4ed3-8fbb-9381bcc3970d" />

Author: Agreed — added Step 1/2/3 subheadings with a one-line description of what each step validates (health check → token generation → full chain). See e006b46.

:keywords: AI, on-premises, getting started, docker, quick start

This guide sets up a fully working local stack in roughly five minutes on any machine with Docker:

* *MySQL 8.0*: conversation history and metadata
* *Redis*: caching and session state
* *TinyMCE AI service*: the on-premises AI back end
* *A minimal token server* (Node.js): signs JWTs for the editor
* *A browser page with TinyMCE*: validates the end-to-end flow

The quick start is designed to validate the stack components before moving to a production deployment. Production engineers can review this section to understand the conceptual flow before continuing to xref:tinymceai-on-premises-production.adoc[Production deployment].

== Before you begin

Verify the following are installed and accessible:

[source,bash]
----
docker --version   # 20.10+ required (or podman --version for Podman 4+)
node --version     # 18+ required
npm --version
----

Additionally, have the following credentials ready:

* A TinyMCE license key and container registry credentials (from the Tiny account representative)
* At least one LLM provider API key (OpenAI, Anthropic, or Google)

> == Five-minute demo with Docker Compose

Reviewer 2: ```suggestion
== Quick demo with Docker Compose
```

Author: Kept as "Getting started" — it's the standard heading for this type of guide and the intro already explains it's a Docker Compose quick start. The page title + first paragraph set the context.

=== Launch the AI service

The AI service runs as a standalone container outside of the Docker Compose stack. This separation allows upgrading or reconfiguring the AI service without restarting the database and Redis.

First, find the Docker Compose network name (Docker creates it from the folder name):

[source,bash]
----
docker network ls --format '{{.Name}}' | grep default
----

Use the matching network name in `--network` below. Then run from the same folder as the `.env` file:

[source,bash]
----
set -a && source .env && set +a

docker run --init -d -p 8000:8000 \
  --network <compose-network-name>_default \
  --name ai-service \
  -e LICENSE_KEY="$LICENSE_KEY" \
  -e ENVIRONMENTS_MANAGEMENT_SECRET_KEY="$MANAGEMENT_SECRET" \
  -e DATABASE_DRIVER='mysql' \
  -e DATABASE_HOST='mysql' \
  -e DATABASE_USER='root' \
  -e DATABASE_PASSWORD="$DB_PASSWORD" \
  -e DATABASE_DATABASE='ai_service' \
  -e REDIS_HOST='redis' \
  -e PROVIDERS='{"openai":{"type":"openai","apiKeys":["'"$OPENAI_API_KEY"'"]}}' \
  -e STORAGE_DRIVER='database' \
  -e ALLOWED_ORIGINS='http://localhost:3000' \
  -e ENABLE_METRIC_LOGS='true' \
  registry.containers.tiny.cloud/ai-service-tiny:latest
----

> TIP: The network name is typically `<foldername>_default` (e.g., `tinymce-ai-onpremise_default`). Run `docker network ls` to confirm. For multiple LLM providers, extend the `PROVIDERS` JSON: `{"openai":{...},"anthropic":{...}}`.

Reviewer 2: ```suggestion
TIP: The network name is typically `<foldername>_default` (e.g., `tinymceai-onpremise_default`). Run `docker network ls` to confirm. For multiple LLM providers, extend the `PROVIDERS` JSON: `{"openai":{...},"anthropic":{...}}`.
```

Reviewer: The folder creation commands in the earlier section are `mkdir tinymce-ai-onpremise && cd tinymce-ai-onpremise` so the original is correct. This is not a TinyMCE option. Just depends on what they called the folder

Author: Applied — updated the TIP to reference the folder name from the mkdir step and added a note about Docker Compose version differences. See 29e0a1c.

[source,bash]
----
curl http://localhost:8000/health
----

Expected response:

[source,json]
----
{"serviceName":"on-premises-http","uptime":5.123}
----

.Successful boot log (`docker logs ai-service`)
[source,text]
----
Connecting to database (driver=mysql host=mysql)
Running migrations...
Migrations complete: 32 tables ready
Connecting to Redis (host=redis:6379)
Redis connected
Server is listening on port 8000.
----

[WARNING]
--
If the container exits immediately, run `docker logs ai-service`. The most common causes are documented in the xref:tinymceai-on-premises-troubleshooting.adoc[Troubleshooting] guide. The top three are: malformed `LICENSE_KEY` (line breaks from word wrap), missing PostgreSQL schema, and JSON syntax error in `PROVIDERS`.
--

=== Create an environment and access key

The AI service isolates users into Environments. Each environment has its own access keys.

. Open the Management Panel: *http://localhost:8000/panel/*
. Sign in using the `MANAGEMENT_SECRET` from `.env`.
. Click *Create Environment* and give it a name (for example "Development").
. Note the *Environment ID* displayed (a short identifier like `viOu8BnjJHb0HGK091p`).

> . Inside the environment, click *Create Access Key*.

Reviewer 2: ```suggestion
. Inside the environment, click *Create a new access key*.
```

Reviewer 2: <img width="1097" height="420" alt="Image" src="https://github.com/user-attachments/assets/950a94eb-f1cb-4e32-a448-8fca37533a3f" />

Author: Applied — updated to "Inside the environment, click *Create a new access key*." See 29e0a1c.

Author: Thanks for the screenshot — helpful for verifying the UI matches our steps.

> . *Copy the API Secret immediately*. The Management Panel shows it only once.

Reviewer 2: ```suggestion
. *Copy Environment ID and Access Key*. 
```

Reviewer 2: Users can copy envid and access key anytime.

<img width="1445" height="554" alt="Image" src="https://github.com/user-attachments/assets/889129ee-bed1-432a-8bdf-ca0b0537a8ce" />

Author: Applied — updated to "Copy the Environment ID and Access Key". See 29e0a1c.

Author: Good to know — thanks for confirming with the screenshot. The step wording now reflects that both can be copied from the panel.

Update `.env` with the new values:

[source,bash]

> ----
> AI_ENV_ID=PASTE_ENVIRONMENT_ID_HERE
> AI_API_SECRET=PASTE_API_SECRET_HERE
> ----

Reviewer 2: ```suggestion
----
AI_ENV_ID=PASTE_ENVIRONMENT_ID_HERE
AI_API_SECRET=PASTE_ACCESS_KEY_HERE
----
```

Author: Kept our standardized `<kebab-case>` placeholder format for consistency across all on-prem pages. The PASTE_X_HERE style was used in the initial draft but we've since normalized everything to angle-bracket descriptors.

Author: Good catch on the inconsistency. Normalized both lines to `<kebab-case>` format in 336cfda.

Kept `<paste-api-secret-here>` rather than `<paste-access-key-here>` so the placeholder matches the `AI_API_SECRET` variable name — avoids confusion about which value goes where.

--

=== Create the `.env` file

[source,bash]
----
# --- Required: AI service license key provided by Tiny ---
AI_LICENCE_KEY=<paste-ai-licence-key-here>

# --- Required for this demo (loads TinyMCE from cdn.tiny.cloud) ---
# Omit only when using a self-hosted editor bundle with license_key.
TINYMCE_API_KEY=<paste-api-key-here>

# --- Required: strong secret used to log into the Management Panel ---
MANAGEMENT_SECRET=<replace-with-strong-secret>

# --- Required: database password (must match docker-compose.yml) ---
DB_PASSWORD=<replace-with-db-password>

# --- Required: at least one LLM provider key ---
OPENAI_API_KEY=<paste-openai-key-here>
# ANTHROPIC_API_KEY=<paste-anthropic-key-here>
# GOOGLE_API_KEY=<paste-google-key-here>

# --- Filled in after creating an environment (see "Create an environment and access key" below). Used by the token server, not the AI service. ---
AI_ENV_ID=<paste-environment-id-here>

> AI_API_SECRET=<paste-api-secret-here>

Reviewer: I had to comment these out to get `set -a && source .env && set +a` to run - kept complaining about newlines at the end of the file (usually means it failed to parse something - in this case, the placeholders).

Author: Addressed in a prior commit — placeholders changed from angle-brackets to single-quoted strings to prevent shell parse errors when sourcing .env.

=== Start MySQL and Redis

[source,bash]
----
docker compose up -d
----

Wait ~15 seconds for MySQL to initialize, then verify:

[source,bash]
----
docker compose ps
----

Both data layer containers (MySQL and Redis) should report `healthy` in the STATUS column. If MySQL still shows `starting`, wait another 10 seconds and re-run.

=== Launch the AI service

The AI service runs as a standalone container outside of the Docker Compose stack. This separation allows upgrading or reconfiguring the AI service without restarting the database and Redis.

First, find the Docker Compose network name (Docker creates it from the folder name):

[source,bash]
----
docker network ls --format '{{.Name}}' | grep default
----

Use the matching network name in `--network` below. Then run from the same folder as the `.env` file:

[source,bash]
----

> set -a && source .env && set +a

Reviewer: fun fact, even if the first line fails, the second could still silently succeed, and then you get a weird situation where you fix the error on the first line, try to rerun this chunk, and get a new error about 'ai-service' already existing. 
maybe we should split these into two snippets?

error about service already existing for context:
<img width="1510" height="327" alt="Image" src="https://github.com/user-attachments/assets/4f0566ee-ad62-494c-9691-a5278dc54ddd" />

Author: Addressed in a prior commit — split into two separate code blocks with explanatory text between them so a failed `source .env` is visible before `docker run` executes.

docker run --init -d -p 8000:8000 \

>   --network <compose-network-name>_default \

Reviewer: I found the first command returns the name with "_default" already appended?

Author: Added bold subheadings (*What this checks:*, *Expected response:*) and brief diagnostic guidance to each of the three verification steps (health check, generate a token, create a conversation). See 74aa678.

Author: Addressed in a prior commit — the TIP was rewritten to explain that `docker network ls` already returns the full name including `_default`, so users copy-paste the name as-is.
